hbase-server/src/main/java/org/apache/hadoop/hbase/mob/DefaultMobStoreCompactor.java

```diff
@@ -183,105 +271,184 @@ protected boolean performCompaction(FileDetails fd, InternalScanner scanner, Cel
+                  LOG.error("Missing MOB cell: file={} not found cell={}", fName, c);
+                  continue;
+                } else {
+                  throw fnfe;
+                }
+              }
+
+              if (discardMobMiss && mobCell.getValueLength() == 0) {
+                LOG.error("Missing MOB cell value: file={} cell={}", pp, mobCell);
+                continue;
+              } else if (mobCell.getValueLength() == 0) {
+                String errMsg = String.format("Found 0 length MOB cell in a file=%s cell=%s",
+                  fName, mobCell);
+                throw new IOException(errMsg);
+              }
+
+              if (mobCell.getValueLength() > mobSizeThreshold) {
+                // put the mob data back to the MOB store file
+                PrivateCellUtil.setSequenceId(mobCell, c.getSequenceId());
+                if (!ioOptimizedMode) {
+                  mobFileWriter.append(mobCell);
+                  mobCells++;
+                  writer.append(
+                    MobUtils.createMobRefCell(mobCell, fileName, this.mobStore.getRefCellTags()));
+                } else {
+                  // I/O optimized mode
+                  // Check if MOB cell origin file size is
+                  // greater than threshold
+                  Long size = mobLengthMap.get().get(fName);
+                  if (size == null) {
+                    // FATAL error, abort compaction
+                    String msg = String.format(
+                      "Found unreferenced MOB file during compaction %s, aborting compaction %s",
+                      fName, getStoreInfo());
+                    throw new IOException(msg);
+                  }
+                  // Can not be null
+                  if (size < maxMobFileSize) {
+                    // If MOB cell origin file is below threshold
+                    // it is get compacted
+                    mobFileWriter.append(mobCell);
+                    // Update number of mobCells in a current mob writer
+                    mobCells++;
+                    writer.append(
+                      MobUtils.createMobRefCell(mobCell, fileName, this.mobStore.getRefCellTags()));
+                    // Update total size of the output (we do not take into account
+                    // file compression yet)
+                    long len = mobFileWriter.getPos();
+                    if (len > maxMobFileSize) {
+                      LOG.debug("Closing output MOB File, length={} file={}, store=", len,
+                        Bytes.toString(fileName), getStoreInfo());
+                      commitOrAbortMobWriter(mobFileWriter, fd.maxSeqId, mobCells, major);
+                      mobFileWriter = newMobWriter(fd);
+                      fileName = Bytes.toBytes(mobFileWriter.getPath().getName());
+                      mobCells = 0;
+                    }
+                  } else {
+                    // We leave large MOB file as is (is not compacted),
```

Comment:
be more specific, plz. Which IO are you referring to?

Reply:
The IO of resolving the MOB value out of the mob hfile.

Reply:
Maybe I am missing something, but I do not see any IO around line 367?

Reply:
I believe this block covers the case where we have a value that we're going to keep in the MOB area, but the mob hfile that currently holds that value is above the threshold that controls if we're continue to have that hfile participate in compactions.

If that's correct, then two things

1) the use of `mobCell` to determine that the value size is over the mob threshold (around line 326) means we resolved the MOB value (looks like line ~307). We could avoid the IO needed to pull that value up out of the backing file, because the size is stored in the reference cell `c`.

2) here in the block we store the name of the backing hfile in the metadata, but we write into the non-mob hfile the whole mob cell rather than the original reference cell `c`. I missed this when I was first thinking about (1) above, but I think that's incorrect?